src/overlays/actors/ovl_Bg_Gnd_Soulmeiro/z_bg_gnd_soulmeiro.c

```diff
@@ -31,17 +31,182 @@ const ActorInit Bg_Gnd_Soulmeiro_InitVars = {
     (ActorFunc)BgGndSoulmeiro_Update,
     (ActorFunc)BgGndSoulmeiro_Draw,
 };
-*/
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Bg_Gnd_Soulmeiro/BgGndSoulmeiro_Init.s")
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Bg_Gnd_Soulmeiro/BgGndSoulmeiro_Destroy.s")
+static ColliderCylinderInit sCylinderInit = {
+    { COLTYPE_UNK10, 0x00, 0x09, 0x00, 0x10, COLSHAPE_CYLINDER },
+    { 0x00, { 0xFFCFFFFF, 0x00, 0x00 }, { 0x00020800, 0x00, 0x00 }, 0x00, 0x01, 0x00 },
+    { 50, 20, 20, { 0, 0, 0 } },
+};
+
+static InitChainEntry sInitChain[] = {
+    ICHAIN_VEC3F_DIV1000(scale, 100, ICHAIN_CONTINUE),
+    ICHAIN_F32(uncullZoneForward, 1000, ICHAIN_CONTINUE),
+    ICHAIN_F32(uncullZoneScale, 1000, ICHAIN_CONTINUE),
+    ICHAIN_F32(uncullZoneDownward, 1000, ICHAIN_STOP),
+};
+
+static Vec3f zeroVec = { 0.0f, 0.0f, 0.0f };
+
+
+Gfx* D_8087B578[] = { 0x06007C00, 0x06002320, 0x060035A0 };
+
+void BgGndSoulmeiro_Init(Actor* thisx, GlobalContext* globalCtx) {
+    BgGndSoulmeiro* this = THIS;
+    s32 pad;
+
+    Actor_ProcessInitChain(&this->actor, sInitChain);
+    this->actionFunc = NULL;
+
+    switch (this->actor.params & 0xFF) {
+        case 0:
+            Collider_InitCylinder(globalCtx, &this->collider);
+            Collider_SetCylinder(globalCtx, &this->collider, &this->actor, &sCylinderInit);
+            this->actionFunc = func_8087B284;
+            if (Flags_GetSwitch(globalCtx, (this->actor.params >> 8) & 0x3F)) {
+
+                Actor_Spawn(&globalCtx->actorCtx, globalCtx, ACTOR_MIR_RAY, this->actor.posRot.pos.x,
+                            this->actor.posRot.pos.y, this->actor.posRot.pos.z, 0, 0, 0, 9);
+                this->actor.draw = NULL;
+                Actor_Kill(&this->actor);
+                return;
+            } else {
+                this->actor.draw = BgGndSoulmeiro_Draw;
+            }
+            break;
+        case 1:
+        case 2:
+            if (Flags_GetSwitch(globalCtx, (this->actor.params >> 8) & 0x3F)) {
+                this->actor.draw = BgGndSoulmeiro_Draw;
+            } else {
+                this->actor.draw = NULL;
+            }
+            this->actionFunc = func_8087B350;
+            break;
+    }
+}
+
+void BgGndSoulmeiro_Destroy(Actor* thisx, GlobalContext* globalCtx) {
+    BgGndSoulmeiro* this = THIS;
+
+    if ((this->actor.params & 0xFF) == 0) {
+        Collider_DestroyCylinder(globalCtx, &this->collider);
+    }
+}
+
+void func_8087AF38(BgGndSoulmeiro* this, GlobalContext* globalCtx) {
+    Vec3f vecA;
+    Vec3f vecB;
+    s16 temp_1;
+    s16 temp_2;
+    f32 temp_3;
+    f32 temp_4;
+    f32 distXZ;
+    s32 i;
+    BgGndSoulmeiro* this2;
+
+    (this->unk_198) == 0 ? 0 : ((this->unk_198--));
+
+    if (this->unk_198 == 20) {
+        Flags_SetSwitch(globalCtx, (this->actor.params >> 8) & 0x3F);
+        this->actor.draw = NULL;
+    }
+
+    // This should be this->unk_198 != 0, this is required to match
+    if (!this->unk_198) {
```

Comment:
Does `== 0` not work? (`!= 0` isn't equivalent to `!this->unk_198`)
```suggestion
    if (this->unk_198 == 0) {
```

Reply:
Unfortunately not, if I remember from my past days when learning very old C, there was a tendency to convert this like var == 0 to !var, even though they did not represent a boolean. In this case the former wouldn't work.

Reply:
In that case I think the comment should be updated at least, so it's not confusing